pkg/controller/core.oam.dev/v1alpha2/application/generator.go

```diff
@@ -83,84 +104,120 @@ func (h *AppHandler) GenerateApplicationSteps(ctx monitorContext.Context,
-		tasks = append(tasks, task)
+	run, err := generateWorkflowRun(af, app, appRev.Name)
+	if err != nil {
+		return nil, nil, err
 	}
-	return tasks, nil
+	executor.InitializeWorkflowRun(run)
+	runners, err := steps.Generate(ctx, run, wfTypes.StepGeneratorOptions{
+		Providers:       handlerProviders,
+		PackageDiscover: h.r.pd,
+		ProcessCtx:      pCtx,
+		TemplateLoader:  template.NewWorkflowStepTemplateRevisionLoader(appRev, h.r.dm),
+		Client:          h.r.Client,
+		StepConvertor: map[string]func(step workflowv1alpha1.WorkflowStep) (workflowv1alpha1.WorkflowStep, error){
+			wfTypes.WorkflowStepTypeApplyComponent: func(lstep workflowv1alpha1.WorkflowStep) (workflowv1alpha1.WorkflowStep, error) {
+				copierStep := lstep.DeepCopy()
+				if err := convertStepProperties(copierStep, app); err != nil {
+					return lstep, errors.WithMessage(err, "convert [apply-component]")
+				}
+				copierStep.Type = wfTypes.WorkflowStepTypeBuiltinApplyComponent
+				return *copierStep, nil
+			},
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return run, runners, nil
 }
 
-func generateStep(ctx context.Context,
-	app *v1beta1.Application,
-	step v1beta1.WorkflowStep,
-	taskDiscover wfTypes.TaskDiscover,
-	pd *packages.PackageDiscover,
-	pCtx process.Context,
-	parentStepName string) (wfTypes.TaskRunner, error) {
-	options := &wfTypes.GeneratorOptions{
-		ID:              generateStepID(step.Name, app.Status.Workflow, parentStepName),
-		PackageDiscover: pd,
-		ProcessContext:  pCtx,
-	}
-	generatorName := step.Type
-	switch {
-	case generatorName == wfTypes.WorkflowStepTypeApplyComponent:
-		generatorName = wfTypes.WorkflowStepTypeBuiltinApplyComponent
-		options.StepConvertor = func(lstep v1beta1.WorkflowStep) (v1beta1.WorkflowStep, error) {
-			copierStep := lstep.DeepCopy()
-			if err := convertStepProperties(copierStep, app); err != nil {
-				return lstep, errors.WithMessage(err, "convert [apply-component]")
-			}
-			return *copierStep, nil
-		}
-	case generatorName == wfTypes.WorkflowStepTypeStepGroup:
-		var subTaskRunners []wfTypes.TaskRunner
-		for _, subStep := range step.SubSteps {
-			workflowStep := v1beta1.WorkflowStep{
-				Name:       subStep.Name,
-				Type:       subStep.Type,
-				Properties: subStep.Properties,
-				DependsOn:  subStep.DependsOn,
-				Inputs:     subStep.Inputs,
-				Outputs:    subStep.Outputs,
-				If:         subStep.If,
-				Timeout:    subStep.Timeout,
-				Meta:       subStep.Meta,
-			}
-			subTask, err := generateStep(ctx, app, workflowStep, taskDiscover, pd, pCtx, step.Name)
-			if err != nil {
-				return nil, err
+func generateWorkflowRun(af *appfile.Appfile, app *v1beta1.Application, appRev string) (*workflowv1alpha1.WorkflowRun, error) {
+	revAndSpecHash, err := workflow.ComputeWorkflowRevisionHash(appRev, app)
+	if err != nil {
+		return nil, err
+	}
+	anno := make(map[string]string)
+	if af.Debug {
+		anno[wfTypes.AnnotationWorkflowRunDebug] = "true"
+	}
+	run := &workflowv1alpha1.WorkflowRun{
```

Comment:
why we need to leverage the workflowRun CRD?

Reply:
We're not using the CRD, just the workflowRun struct to use to workflow engine.